### DISCUSSION TITLE
Errors after ending response in getServerSideProps

### DISCUSSION THREAD

**Discussion by cbrianball** [vercel/next.js#35549] [Help] [answered]
Under certain conditions, getServerSideProps, in my app, can fully handle the response, and therefore the page does not need to be rendered.

I am unable to do this in middleware as I need access to native nodejs modules in order to determine if the request can be fulfilled by the page without needing to render.

Using /api is possible (and I do use it in my work around), but the request to the resource is much cleaner without the API path (e.g., /path/to/some/file.txt instead of /api/download/path/to/some/file.txt). My work around involves returning a redirect from my getServerSideProps method to the API URL, but I would much rather not have the redirect.

If I let getServerSideProps handle it (and end the response manually), from the user's perspective, everything works, but the server logs are showing errors of attempting to write the header after the headers have already been sent). I assume this is because I've already written the body and next.js is trying to write headers for the page that it is about to render.

**Comment by icyJoseph** [comment, marked as answer]
> I am unable to do this in middleware as I need access to native nodejs modules in order to determine if the request can be fulfilled by the page without needing to render.

Which API, exactly, do you need?

> Under certain conditions, getServerSideProps, in my app, can fully handle the response, and therefore the page does not need to be rendered.

> If I let getServerSideProps handle it (and end the response manually), from the user's perspective, everything works, but the server logs are showing errors of attempting to write the header after the headers have already been sent). I assume this is because I've already written the body and next.js is trying to write headers for the page that it is about to render.

What I gather is that you are breaking `getServerSideProps` behaviour. The return type of `getServerSideProps` has to be `{redirect:{destination: '/somewhere'}}, `{props:{/* some props */}`, or `{notFound: boolean}`, though you can of course work with `ctx.res`, to set cookies, and other stuff proper.

Though I wonder, what happens if you do something like (against my better judgement)

```js
const Post = () => <div>Post</div>;

export const getServerSideProps = (ctx) => {
  const path = new URL("", ctx.req.headers.referer); 
  return { redirect: { destination: path.pathname, statusCode: 302 } };
};

export default Post;
```

How should your use case work if I navigate to your resource directly? I'd see nothing at all? I think you do need to have an `/api` route to check if you should allow the navigation.

**Reply by cbrianball** [reply]
I need to use `fs`. My site is a web view into the NAS on my network that I can access from anywhere. I could've found an existing solution, but it gives me an opportunity to try out new [to me] frameworks, like next.js.

In cases where `getServerSideProps` is given a path that points to a directory, the page is returned, in HTML, for the user to view the contents of the folder. If the path points to a file, then the result should be the file itself

If I'm using `getServerSideProps` in a manner that wasn't intended, that's fine, my current workaround of having it send a redirect to an `/api` endpoint functions just fine. I'm not sure if this would make sense, but next.js could check to see if the response has ended after the call to `getServerSideProps` and, if so, essentially do nothing (trust that it has already been taken care of).

Normally this would be a perfect fit for middleware to make a decision, and if it is a file, issue a rewrite so that it is completely seamless to the browser, but my understanding is that I can't use modules like `fs` since, under certain deployments, the middleware is hosted elsewhere.

The only thing that would make my current work-around better is if `getServerSideProps` could do a rewrite and not a redirect, but based on what little knowledge I have, I suspect it's "too late" to do a rewrite at that point in time.

**Reply by icyJoseph** [reply]
Aha, now I understand, sorry if I was a bit dry on my answer.

Would it be okay to render the contents of a file inside an iframe? If you detect that it is a `file` path name you return in the props `url` to be used by the iframe to "download" the file contents to the client. 

Curious on how are you planning to handle binaries? If the files are only text based, you could read them and send the contents in the `getServerSideProps` response, and have conditional rendering on the client. 

Another thing would be, to check the extensions (?), or to include more info when you analyse a directory tree. I mean for example if you have `/documents`, and it has `/a`, and `/b.txt` to create a model:

```js
const current = {
  path: "/documents",
  contents: [
    { path: "/a", contents: [], type: 'dir' },
    { path: "/b.txt", contents: [], type: 'file' },
  ],
  type: 'dir',
};
```

Sort of putting the calculation upfront for depth 1 in the current directory. Edited to use `type` instead of `isFile`.

With this you could do the check in the middleware, by including the `type` as query/payload in the request, and do a rewrite for `file` types. A bit dirty solution, but perhaps you can polish it. Sadly direct navigation to a resource still needs the redirect. Interesting use case.

**Reply by icyJoseph** [reply]
Ah, sorry for spamming, I couldn't stop thinking of this use case, and I just had an idea:

```ts
import type { NextRequest } from "next/server";
import { NextResponse } from "next/server";

const shouldExclude = (pathname: string) => {
  return (
    pathname.startsWith("/api") || //  exclude all API routes
    pathname.startsWith("/static") || // exclude static files
    pathname.includes(".") // exclude all files in the public folder
  );
};

export async function middleware(req: NextRequest) {
  const { pathname } = req.nextUrl;

  if (shouldExclude(pathname)) return NextResponse.next();

  try {
    const fsCheck = new URL("/api/check_fs", req.nextUrl);
    // basically make a request to your own server to check the file system in a lambda
    const res = await fetch(fsCheck.toString(), {
      method: "POST",
      body: JSON.stringify({
        /* data needed for check_fs */
      }),
      headers: { "Content-Type": "application/json" },
    });

    const result = await res.json();

    // do something based on result
    if (result.type === "file") {
      req.nextUrl.pathname = "/api/see-file";
      return NextResponse.rewrite(req.nextUrl);
    }
    return NextResponse.next();
  } catch (err) {
    return new Response(null, {
      status: 500,
      statusText: "Internal Server Error",
    });
  }
}

```

What do you think?

**Reply by cbrianball** [reply]
I like the idea of having the middleware make a call to the `/api`, then decide what to do; I'll have to give that some thought. It would obviate the need for a redirect and it would ensure that only requests that will result in a webpage being returned would actually make it to the appropriate page itself.

Is there a way of effectively making a specific `/api` endpoint internal (so no one could directly call it from their browser)? If not, it's no big deal, just curious.

As for how I'm handling binaries: I'm going to have a mapping of mime types somewhere that determines if the file should be viewed directly in the browser or downloaded. Files like images, text documents, PDFs, etc. I'm planning to show directly in the browser, other file types will be downloaded.

I'm even planning to handle office documents in a way that will open them in the correct application and saves will automatically write back to the NAS (I had this working in a previous iteration of the site, but next.js really piqued my interest, so I wanted to try it out).

**Reply by icyJoseph** [reply]
> Is there a way of effectively making a specific /api endpoint internal (so no one could directly call it from their browser)? If not, it's no big deal, just curious.

Yes, but in this case, it looks a bit daunting, because you prevent access to the endpoint, but down below, actually access it 😅 

```ts
import { NextRequest, NextResponse } from "next/server";

const shouldExclude = (pathname: string) => {
  return (
    pathname.startsWith("/api") || //  exclude all API routes
    pathname.startsWith("/static") || // exclude static files
    pathname.includes(".") // exclude all files in the public folder
  );
};

export const middleware = async (req: NextRequest) => {
  const { pathname } = req.nextUrl;

  // prevent direct access
  if (pathname.startsWith("/api/check_fs")) {
    return new Response(null, { status: 404, statusText: "Not Found" });
  }

  if (shouldExclude(pathname)) return NextResponse.next();

  const fsCheck = new URL("/api/check_fs", req.nextUrl);
  // basically make a request to your own server to check the file system in a lambda
  const res = await fetch(fsCheck.toString(), {
    method: "POST",
    body: JSON.stringify({
      /* data needed for check_fs */
    }),
    headers: { "Content-Type": "application/json" },
  });

  const result = await res.json();

  // do something based on result
  if (result.type === "file") {
    req.nextUrl.pathname = "/api/see-file";
    return NextResponse.rewrite(req.nextUrl);
  }
  return NextResponse.next();
};
```

It all hangs by a thread, in this case `api/check_fs` would get a header `'x-middleware-subrequest': 'pages/_middleware'`, which I think is responsible for allowing the request. 

If you make a postman request with that header, you'll reach the unreachable! 😅  - A token might be better if security is a concern. I can recommend `jose`, which works fine in `_middleware` environments.

---

> As for how I'm handling binaries: I'm going to have a mapping of mime types somewhere that determines if the file should be viewed directly in the browser or downloaded. Files like images, text documents, PDFs, etc. I'm planning to show directly in the browser, other file types will be downloaded.

> I'm even planning to handle office documents in a way that will open them in the correct application and saves will automatically write back to the NAS (I had this working in a previous iteration of the site, but next.js really piqued my interest, so I wanted to try it out).

This sounds like a really cool project, if you have it open source I'd like to play around with it, I've been thinking of installing a NAS at home for while as well.

**Reply by cbrianball** [reply]
Currently just a private project for me. If I open source it, I'll let you know.

Thanks for all your help!
